src/index/ivf_raft/ivf_raft.cuh

```diff
@@ -0,0 +1,385 @@
+/**
+ * SPDX-FileCopyrightText: Copyright (c) 2023,NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef IVF_RAFT_CUH
+#define IVF_RAFT_CUH
+
+#include <cstddef>
+#include <cstdint>
+#include <optional>
+#include <memory>
+#include "common/raft_metric.h"
+#include "raft/core/device_resources.hpp"
+#include "raft/neighbors/ivf_flat.cuh"
+#include "raft/neighbors/ivf_flat_types.hpp"
+#include "raft/neighbors/ivf_pq.cuh"
+#include "raft/neighbors/ivf_pq_types.hpp"
+#include "thrust/sequence.h"
+#include "thrust/execution_policy.h"
+#include "index/ivf_raft/ivf_raft_config.h"
+#include "knowhere/factory.h"
+#include "knowhere/index_node_thread_pool_wrapper.h"
+
+namespace knowhere {
+
+namespace detail {
+using raft_ivf_flat_index = raft::neighbors::ivf_flat::index<float, std::int64_t>;
+using raft_ivf_pq_index = raft::neighbors::ivf_pq::index<std::int64_t>;
+
+// TODO(wphicks): Replace this with version from RAFT once merged
+struct device_setter {
+    device_setter(int new_device) : prev_device_{[]() {
+        auto result = int{};
+        RAFT_CUDA_TRY(cudaGetDevice(&result));
+        return result;
+    }()} {
+        RAFT_CUDA_TRY(cudaSetDevice(new_device));
+    }
+
+    ~device_setter() {
+        RAFT_CUDA_TRY_NO_THROW(cudaSetDevice(prev_device_));
+    }
+    private:
+    int prev_device_;
+};
+
+}
+
+template <typename T>
+    struct KnowhereConfigType {};
+
+template <>
+    struct KnowhereConfigType<detail::raft_ivf_flat_index> {
+        typedef RaftIvfFlatConfig Type;
+    };
+template <>
+    struct KnowhereConfigType<detail::raft_ivf_pq_index> {
+        typedef RaftIvfPqConfig Type;
+    };
+
+template <typename T>
+    class RaftIvfIndexNode : public IndexNode {
+        public:
+            RaftIvfIndexNode(const Object& object) : devs_{}, res_{std::make_unique<raft::device_resources>()}, gpu_index_{} {}
```

Comment:
maybe we need a device memory pool.

Reply:
This is fairly trivial to add thanks to RMM, but we may need some further discussion about how exactly it should behave with respect to the rest of Knowhere. For instance, it is easy to set up a pool that grows to accommodate however much memory is required, but do we need to set up a mechanism to release that memory if other index types (which do not use RMM) are used in combination with an IVF RAFT index? Alternatively, do we want to have a fix-sized memory pool, and if so how should users pass the desired size?

I'm happy to add the simplest version of a memory pool right away to this PR, but if there are more complex considerations, we might want to look at them in a follow-up.

Reply:
BTW- the explicit rmm memory resource argument for searching is going away soon. However, an rmm memory pool can still be set as the `default` on the current process. The `raft::device_resources` (which is used by all of RAFT's APIs) also contains a `workspace_resource` which is an RMM resource that can be used to override the default RMM resource (for example, for setting a pool memory resource only for temporary device allocations).

Reply:
I've just added basic usage of RMM's pool resource, and we can refine from there.